### ISSUE TITLE
UPowerWidget does not deregister hooks on config reload

### ISSUE THREAD

**Issue by clotodex** [elParaguayo/qtile-extras#92]
when reloading the config, the internal async callbacks are being called multiple times (from previous config and from the new reload).
While this has no impact on functionality it essentially has a "leak" which is not desired.

Further overriding the update function to notify-send will get trigged multiple times then :P

**Comment by elParaguayo** [comment]
Thanks for the report. Will get this fixed.

**Comment by elParaguayo** [comment]
@clotodex I've not had a chance to investigate yet but could you let me know which callbacks are being called multiple times?

If it's `_config_async`, I don't think that really matters for this widget and the issue is really in the qtile code (i.e. I don't think there's anything specific to this widget that means that code gets called multiple times).

**Comment by clotodex** [comment]
@elParaguayo I used ` async def _update_battery_info(self, draw=True):`
calling super() on it and the reading out the info to check if I should notify.
Every time I reload the config, it is getting called one more time.

What I see happening is that when the widget starts up it registers hooks in dbus (not qtile).
It seems qtile assumes widgets to be *well behaved* and thus does not aggressively clean them up (like including async hooks).
So I think you need to somehow deregister them when qtile reloads its config, or not reregister them.

**Comment by elParaguayo** [comment]
It registers hooks in dbus but, when you reload the config, I would expect the old bus objects to be destroyed. Multiple calls on reload is not unexpected because of how the bar configures widgets but if every subsequent update resulted in multiple notifications then I'd agree that that's a problem. Is that what you're seeing?

**Comment by elParaguayo** [comment]
If it is a problem, we can call `disconnect` on the bus object when the widget is finalised.

**Comment by clotodex** [comment]
> It registers hooks in dbus but, when you reload the config, I would expect the old bus objects to be destroyed. Multiple calls on reload is not unexpected because of how the bar configures widgets but if every subsequent update resulted in multiple notifications then I'd agree that that's a problem. Is that what you're seeing?

Yes that is exactly it.
I solved my battery notifier now as its own daemon, so  I am not dependent on the widget anymore. However fixing it would be nice anyways right?
Also if you think the destruction of the widgets is not handled correctly by qtile it might make sense to create an issue over there. @elParaguayo

**Comment by elParaguayo** [comment]
Widgets have a mechanism for destruction (you just add a `finalize` method). I had assumed (incorrectly, it seems) that the old connection would be destroyed as a new `self.bus` object is created when the widget is reconfigured.

Should be an easy fix.

**Comment by elParaguayo** [comment]
could you try adding this method to your class:
``` python
def finalize(self):
    self.bus.disconnect()
    self.bus = None
   # do a super() call here too to make sure the main finalize method is called
```

There is a method to remove signal handlers but it's not accessible with my current code (I'd need to make the properties interface a class attribute). Easy enough to do but the above may be enough for you for now.

**Comment by elParaguayo** [comment]
Can you try the fix in #96?

**Comment by clotodex** [comment]
@elParaguayo I can confirm - it fixed it (sorry for the late reply)

for reference: here my hacky notifier :P
```
class MyUPowerWidget(extrawidget.UPowerWidget):
    async def _update_battery_info(self, draw=True):
        await super()._update_battery_info(draw=draw)
        for battery in self.batteries:
            if int(battery["percentage"]) < 10:
                subprocess.Popen(
                    ["dunstify", "battery", f"low {battery['percentage']} %"]
                )

'''
After #96 this only produces one notification, even when reloading the config - which is the expected behavior.

Thank you for fixing this!
